src/Microsoft.AspNet.Mvc.Core/Routing/UrlHelper.cs

```diff
@@ -142,15 +138,59 @@ protected virtual string GeneratePathFromRoute(string routeName, RouteValueDicti
 
             // VirtualPathData.VirtualPath returns string.Empty for null.
             Debug.Assert(pathData.VirtualPath != null);
-
-            var fullPath = HttpContext.Request.PathBase.Add(pathData.VirtualPath).Value;
-            if (fullPath.Length == 0)
+            var pathBase = HttpContext.Request.PathBase;
+            if (!pathBase.HasValue)
             {
-                return "/";
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    return "/";
+                }
+                else
+                {
+                    return !pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal)
+                        ? "/" + pathData.VirtualPath
+                        : pathData.VirtualPath;
+                }
             }
             else
             {
-                return fullPath;
+                if (pathData.VirtualPath.Length == 0)
+                {
+                    return pathBase;
+                }
+                else
+                {
+                    var requiresLeadingSlash = !pathData.VirtualPath.StartsWith("/", StringComparison.Ordinal);
+                    var virtualPathStartIndex = 0;
+                    if (pathBase.Value.EndsWith("/", StringComparison.Ordinal))
```

Comment:
`PathString.Add` checked for this condition - https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Abstractions/PathString.cs#L192-L193